### PR TITLE
[GGUF STEP4] fix search bug and improve packing & eval speed

### PR DESCRIPTION
**Pull request by n1ck-guo**
TODO:
- [x] fix search bug
- [x] increase packing speed
- [x] increase eval speed
- [ ] phi4 large accuracy drop after converting to HF format while fake model is fine , origin gguf  has similar issue
- [ ] falcon3 2% accuracy drop comparing to fake model
- [ ] q2_k_s large accuracy drop  after converting to HF format, while fake model is fine, origin gguf has simialr issue

### CODE REVIEW

File: auto_round/export/export_to_gguf/quant.py [hunk 1]
```diff
@@ -173,47 +234,46 @@ def make_qkx2_quants(data, weight, nmax, group_size, rmin=-1, rdelta=0.1, nstep=
...
 
     if nstep < 1:
         the_min = -group_min
         return scale, L, the_min
 
-    Laux = []
     for step in range(nstep):
-        iscale = (rmin + rdelta*step + nmax) / (group_max-group_min)
-        l_values = np.round(iscale * (data-group_min))
+        iscale = (rmin + rdelta * step + nmax) / (group_max - group_min)
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant.py]: @SinpackKonmakan  please also try 
1 rmin=0
2 rmin=-2, nstep 40

File: auto_round/utils.py [hunk 2]
```diff
@@ -1185,6 +1185,13 @@ def _gguf_args_check(args):
     pre_dq_format = ""
     for format in GGUF_CONFIG:
         if format in formats:
+            try:
+                from auto_round.export.export_to_gguf.convert import Model
+            except:
+                logger.error(
```
Comment (wenhuach21) [inline auto_round/utils.py]: Raise import error?

File: auto_round/utils.py [hunk 3]
```diff
@@ -1188,8 +1188,8 @@ def _gguf_args_check(args):
             try:
                 from auto_round.export.export_to_gguf.convert import Model
             except:
-                logger.error(
-                    f"Please use the latest gguf-py for {format}, you can use the following command to install it:\n"
+                raise ImportError(
+                    f"Please use the latest gguf-py for {format}, you can use the following command to insall it:\n"
```
Comment (wenhuach21) [inline auto_round/utils.py]: insall->install

**Comment by wenhuach21** [comment]
if output_dir is already existed, logger a warning

**Review by wenhuach21 — APPROVED** [review]

**Comment by wenhuach21** [comment]
test 3-4 models and make sure all the accuracies is reasonable without autoround tuning

File: auto_round/export/export_to_gguf/quant.py [hunk 4]
```diff
@@ -165,216 +167,181 @@ def q8_0_quant_block(blocks: np.array, scale=None, zp=None) -> np.ndarray:
 
     return np.concatenate([d, qs], axis=1)
 
+def make_qkx2_quants(data, bits, rmin=-1, rdelta=0.1, nstep=20, use_mad=False):
+    assert bits in [2, 4], f"bits={bits} is not supported"
+    nmax = bits ** 2 -1
+    # data shape (nb, 8, 32) for Q4_K, (nb, 16, 16) for Q2_K
+    if len(data.shape) == 2:
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant.py:182]: assert bits==2 or 4

Comment (wenhuach21) [reply auto_round/export/export_to_gguf/quant.py:182]: For other bits that use asymmetric quantization, we should add support for them later.

File: auto_round/export/export_to_gguf/quant_cuda.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    def register(cls):
+        GGML_QUANT_BLOCK[name] = cls
+        return cls
+
+    return register
+
+
+def ggml_quant_cuda(data, ggml_type, scale=None, zp=None, wmin_m=None, d_scale=None, d_wmin_m=None):
+    block_size, type_size = GGML_QUANT_SIZES[ggml_type]
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant_cuda.py]: I think using wmin is enough, no need to use zp. But I am not sure

File: auto_round/export/export_to_gguf/quant_cuda.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    floored = torch.floor(a)
+    b = floored + torch.floor(2 * (a-floored))
+    return torch.sign(n) * b
+
+
+@register_block("bf16")
+def bf16_quant_block(blocks, scale=None, zp=None, **kwargs):
+    n = blocks.cpu().numpy().view(np.uint32)
+    # force nan to quiet
+    n = np.where((n & 0x7fffffff) > 0x7f800000, (n & np.uint32(0xffff0000)) | np.uint32(64 << 16), n)
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant_cuda.py]: why theses ops are still on cpu

File: auto_round/export/export_to_gguf/quant_cuda.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+    return register
+
+
+def ggml_quant_cuda(data, ggml_type, scale=None, zp=None, wmin_m=None, d_scale=None, d_wmin_m=None):
+    block_size, type_size = GGML_QUANT_SIZES[ggml_type]
+
+    data = data.to(torch.float32).to("cuda")
+    scale = scale.to("cuda") if scale is not None else scale
+    zp = zp.to("cuda") if zp is not None else zp
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant_cuda.py]: better not  code like this.  Use `device`  arg to set the device, then we could easily extend to xpu and other devices

File: auto_round/export/export_to_gguf/quant_cuda.py (hunk 5 above)
Comment (n1ck-guo) [reply auto_round/export/export_to_gguf/quant_cuda.py]: zp is for some other quant method as q4_1

File: auto_round/script/llm.py [hunk 8]
```diff
@@ -572,15 +578,43 @@ def _eval_init(tasks, model_path, device, disable_trust_remote_code=False):
...
+        for file in os.listdir(model):
+            if file.endswith(".gguf"):
+                is_gguf_file = True
+                gguf_file = file
+    if is_gguf_file:
+        import torch
+        from transformers import AutoTokenizer, AutoModelForCausalLM
+        from lm_eval.utils import make_table  # pylint: disable=E0401
+        tokenizer = AutoTokenizer.from_pretrained(model, gguf_file=gguf_file)
+        user_model = AutoModelForCausalLM.from_pretrained(model, gguf_file=gguf_file, device_map="auto")
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: fix it if the model is mapped to multi-gpus, multi-gpu is controlled by --devices

File: auto_round/export/export_to_gguf/quant_cuda.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+def register_block(name):
+
+    def register(cls):
+        GGML_QUANT_BLOCK[name] = cls
+        return cls
+
+    return register
+
+
+def ggml_quant_cuda(data, ggml_type, scale=None, zp=None, wmin_m=None, d_scale=None, d_wmin_m=None, device="cuda"):
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/quant_cuda.py]: rename the file and function to replace `cuda` with `gpu`

File: auto_round/script/llm.py [hunk 10]
```diff
@@ -192,6 +192,8 @@ def __init__(self, *args, **kwargs):
         self.add_argument(
             "--super_bits", default=None, type=int, help="number of scale and mins quant bits for double quant.")
 
+        self.add_argument("--torch_dtype", default=None, type=str, help="torch_dytpe to load the model.")
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: change to another name eval_model_dtype? As we have a model_dtype in args,  and change the discription

File: auto_round/script/llm.py [hunk 11]
```diff
@@ -525,8 +528,25 @@ def tune(args):
...
+                logger.warning(f"Unable to identify {torch_dtype}, reset to float32")
+
+            if torch_dtype == torch.float32:
+                logger.warning(
+                    "set '--torch_dtype bf16' can significantly speed up evaluation for gguf model,"
+                    " but may affect accuracy."
+                )
             model = AutoModelForCausalLM.from_pretrained(
-                eval_folder, gguf_file=gguf_file, device_map="auto", torch_dtype=torch.bfloat16)
+                eval_folder, gguf_file=gguf_file, device_map="auto", torch_dtype=torch_dtype)
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: better apply to other formats to keep consistent

File: auto_round/script/llm.py [hunk 12]
```diff
@@ -671,7 +734,8 @@ def eval_task_by_task(
     for task in tasks:
         while retry_times:
             try:
-                res = lm_simple_evaluate(model=hflm, model_args=None, device=device_str, tasks=task, batch_size=batch_size)
+                res = lm_simple_evaluate(
+                    model=hflm, model_args=None, device=device_str, tasks=task, batch_size=batch_size, limit=500)
```
Comment (wenhuach21) [inline auto_round/script/llm.py]: rm limit
